File: EASY/Easy_Cpp/word_pattern.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    bool wordPattern(string pattern, string s) {
        vector<string> words;
        string word;
        for (char c : s) {
            if (c == ' ') {
                words.push_back(word);
                word = "";
            }
            else {
                word += c;
            }
        }
        words.push_back(word);

        if (pattern.size() != words.size()) {
            return false;
        }

        unordered_map<char, string> charToWord;
        unordered_map<string, char> wordToChar;

        for (int i = 0; i < pattern.size(); i++) {
            char c = pattern[i];
            string w = words[i];
            if (charToWord.count(c) && charToWord[c] != w) {
                return false;
            }
            if (wordToChar.count(w) && wordToChar[w] != c) {
                return false;
            }
            charToWord[c] = w;
            wordToChar[w] = c;
        }

        return true;
    }
};

File: EASY/word_pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split(" ")

        # check if it's valid
        if len(pattern) != len(words):
            return False
        
        charToWord = {} # hash map
        wordToChar = {} # hash map

        for c, w in zip(pattern, words):
            if c in charToWord and charToWord[c] != w:
                return False

            if w in wordToChar and wordToChar[w] != c:
                return False
        
            charToWord[c] = w
            wordToChar[w] = c
            
        return True
